### PR TITLE
231 establish table of active NPIs according to Corona Datenplattform and reported infection data

### LINKED ISSUE #231: Transformation of NPI data and preliminary analysis
The https://www.corona-datenplattform.de/ provides finely granulated information on non-pharmaceutical interventions (NPIs) against Sars-CoV-2 on county scale. This information is extracted from official political publications and NPIs are mentioned when they were considered for implementation. However, for incidence-dependent NPIs a combination with incidence data has to done to get to know when which NPIs were active (and not only mentioned in the regulations).

In a first step, this information has to be brought in a pandas data frame as our other data. After combination with incidence data, clustering may be needed to cluster interventions that are highly correlated with respect to the dates when they were active.

### LINKED ISSUE #412: Compare data on activation of NPIs from different sources
Compare data on activation of NPIs from https://www.corona-datenplattform.de/ with data from 'Amtsblättern' for Cologne

### LINKED ISSUE #413: Compare old and new NPI data
Check if new and old data for NPIs coincide

### LINKED ISSUE #414: Check transformNPIData
Go through details of function transform_npi_data on branch 231-npi-data

### PR DESCRIPTION
**Pull request by mknaranja**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [x] Tests for new functionality has been added
* [x] A local test was succesful
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)
* [ ] If new third party software is used, did you pay attention to its license? Please remember to add it to the wiki after successful merging.
* [ ] If new mathematical methods or epidemiological terms are used, has the glossary been updated ? Did you provide further documentation ?
 is present or referenced. Please provide your references.
* [ ] The following questions are addressed in the documentation*:  Developers (what did you do?, how can it be maintained?), For users (how to use your work?), For admins (how to install and configure your work?)
* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [x] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [x] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [x] There are appropriate **unit tests** and they pass.
* [x] The git history is clean and linearized for the merge request.
* [x] Coverage report for new code is acceptable. 

Closes #231
Closes #412
Closes #413
Closes #414

### CODE REVIEW

File: pycode/memilio-epidata/memilio/epidata/transformNPIData.py [hunk 1]
```diff
@@ -0,0 +1,1339 @@
...
+                    # int_active should then be:
+                    # [0,1,1,1,1,1,1,1,1,1,1,1,0]
+
+                    # get a zero filled Series with same length to be
+                    # filled with ones where NPI is active
+                    int_active = pd.Series(np.zeros(len(local_incid), dtype=int))
+                    # loop over every day
+                    for i in range(len(yesterdays_incid_over_threshold)):
+                        # Set int_active=0 where last npi_lifting_delay+1 days are 0
+                        if yesterdays_incid_over_threshold[i-npi_lifting_delay:i+1].values.sum() == 0:
```
Comment (mknaranja) [inline pycode/memilio-epidata/memilio/epidata/transformNPIData.py:871]: @patricklnz Why is this running until i+1?

File: pycode/memilio-epidata/memilio/epidata/transformNPIData.py [hunk 2]
```diff
@@ -551,13 +558,14 @@ def transform_npi_data(fine_resolution=2,
         # extract variable names for main categories
         npi_desc = list(df_npis_desc["Variable"][npi_codes_sorting])
 
+    del df_npis_desc
+
     # combine NPI codes and descriptions to ensure that both are ordered
     # the same way; delete npi_codes or npi_desc for not using hereafter
     idx_codes_retained = ~pd.Series(npi_codes).isin(codes_dropped)
-    npis_dummy = {
+    npis = pd.DataFrame({
```
Comment (mknaranja) [inline pycode/memilio-epidata/memilio/epidata/transformNPIData.py:566]: why renaming this npis and deleting afterwards? Could just have been npis_dummy...

**Comment by mknaranja** [comment]
analysis extracted; see #444 @patricklnz

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 3]
```diff
@@ -1039,40 +1037,22 @@ def get_npi_data(fine_resolution=2,
         pass
 
     #### start validation ####
-    if fine_resolution == 2 and (npi_activation_days_threshold + npi_lifting_days_threshold == 0):
-        start_date_validation = datetime(2020, 3, 1)
-        end_date_validation = datetime(2022, 2, 15)
+    if fine_resolution == 2 and npi_activation_days_threshold == 1 and npi_lifting_days_threshold == 1 or fine_resolution == 1:
```
Comment (mknaranja) [inline pycode/memilio-epidata/memilio/epidata/getNPIData.py:1040]:
```suggestion
    if fine_resolution >0 and npi_activation_days_threshold == 1 and npi_lifting_days_threshold == 1:
```

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 4]
```diff
@@ -1006,117 +1015,68 @@ def get_npi_data(fine_resolution=2,
...
+                for i in range(len(idx_strictness_sorted_rev)-1):
                     # get index of NPI of a certain strictness
-                    idx_strictness = idx_strictness_sorted_rev[jj]
+                    idx_strictness = idx_strictness_sorted_rev[i]
                     # get code of corresponding NPI
-                    subcode = subcodes_list[idx_strictness]
+                    subcode = subcodes[idx_strictness]
+
+                    # get subcode index
+                    scidx = subcodes.index(subcode)
```
Comment (mknaranja) [inline pycode/memilio-epidata/memilio/epidata/getNPIData.py:1055]: scidx = idx_strictness?

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 5]
```diff
@@ -1006,117 +1015,68 @@ def get_npi_data(fine_resolution=2,
...
 
                     # get indices of days where subcode is active
-                    subcode_active = np.where(df_local_new.loc[:,subcode]>0)[0]
+                    subcode_active = np.where(df_merged.loc[:,subcode]>0)[0]
 
-                    if len(subcode_active>0):                  
+                    if len(subcode_active) > 0:                  
                         # get indices of less strict NPIs
-                        idxs_less_strict = np.sort(idx_strictness_sorted_rev[jj+1:])
+                        idxs_less_strict = df_npis_combinations[maincode][1].columns[np.sort(idx_strictness_sorted_rev[i+1:])]
```
Comment (mknaranja) [inline pycode/memilio-epidata/memilio/epidata/getNPIData.py:1062]: with df_npis_combinations[maincode][1].columns[] it is not an index anymore

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 6]
```diff
@@ -1006,117 +1015,68 @@ def get_npi_data(fine_resolution=2,
...
                         # only consider those codes which cannot be combined; 
                         # for these, values of 1 have to be set to 0
                         subcodes_nocombi = list(
                             subcodes_nocombi
                             [subcodes_nocombi == 0].index)
                         
                         # intersect non-combinable subcodes with less strict subcodes
-                        idx_subcodes_deactivation = np.sort(list(
-                            set(idxs_less_strict).intersection(subcodes_nocombi)))
+                        subcodes_deactivation = list(set(idxs_less_strict).intersection(subcodes_nocombi))
```
Comment (mknaranja) [inline pycode/memilio-epidata/memilio/epidata/getNPIData.py:1072]: stopped here

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 7]
```diff
@@ -702,9 +717,7 @@ def get_npi_data(fine_resolution=2,
                 if key in npis['NPI_code'].values}
             # remove columns of combinations
             df_npis_combinations[code][1] = df_npis_combinations[code][1].loc[local_codes_used_rows,
-                                                                           local_codes_used_cols].reset_index(drop=True).copy()
-
-
+                                                                              local_codes_used_cols].reset_index(drop=True).copy()
```
Comment (mknaranja) [inline pycode/memilio-epidata/memilio/epidata/getNPIData.py:720]: simple shallow copy() ok here?

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 8]
```diff
@@ -1006,117 +1053,80 @@ def get_npi_data(fine_resolution=2,
...
-            # Remove conflicting non-incidence (?) dependent NPIs according to
-            # strictness index of Corona-Datenplattform and exclusion
-            # criteria defined in df_npis_combinations
-            for code in df_npis_combinations.keys():
-                # get other subcodes MX_Y_* below current main code MX_Y
-                subcodes_list = list(df_npis_combinations[code][0].keys())
+            # merge incidence dependent NPIs to have only one column for each subcode
+            df_merged = df_local_new.iloc[:, :2].copy()
+            for subcode in all_subcodes:
+                df_merged[subcode] = df_local_new.filter(
```
Comment (mknaranja) [inline pycode/memilio-epidata/memilio/epidata/getNPIData.py:1059]: probably filter is the better technique than str, but why didn't you actually use this filtering in `count_codes` in `df_local[df_local.NPI_code.str.contains(column)]`?

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 9]
```diff
@@ -1196,6 +1216,133 @@ def get_npi_data(fine_resolution=2,
...
+
+    return df_count_incid_depend
+
+
+def save_counter(df_count, filename, directory):
+    # save results
+
+    writer = pd.ExcelWriter(
+        os.path.join(directory, filename + '.xlsx'),
+        engine='xlsxwriter')
```
Comment (mknaranja) [inline pycode/memilio-epidata/memilio/epidata/getNPIData.py:1327]: `xlsxwriter` needs to be added to install_requires https://github.com/DLR-SC/memilio/blob/main/pycode/memilio-epidata/setup.py#L73

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 10]
```diff
@@ -895,39 +942,39 @@ def get_npi_data(fine_resolution=2,
...
+                                                 == countyID])
 
         # Consistency of incidence dependent NPIs:
         # The same NPI should not be prescribed multiple times at the same day
         # for different thresholds. In order to avoid contradictions, only
         # retain the strictest mentioned implementation.
         for i in range(int(len(df_local_old)/6)):
             sum_npi_inc = np.where(
-                df_local_old.iloc[6*i+1:6*(i+1), 6:].sum() > 1)
+                df_local_old.iloc[6*i:6*(i+1), 6:].sum() > 1)
```
Comment (mknaranja) [inline pycode/memilio-epidata/memilio/epidata/getNPIData.py:962]: we should think of replacing the two 6's by different variables. The first represents the number of lines per NPI, the second the start of the NPI columns. Also applies to other occasions.

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 11]
```diff
@@ -1167,47 +1215,134 @@ def get_npi_data(fine_resolution=2,
...
             for column in range(len(column_list)):
-                for column_other in range(len(column_list)):
-                    df_npis_combinations[code][1].iloc[column, column_other] += len(set(code_dict[column_list[column]]).intersection(set(code_dict[column_list[column_other]])))
-                    df_npis_combinations[code][1].iloc[column_other, column] += len(set(code_dict[column_list[column_other]]).intersection(set(code_dict[column_list[column]])))
-
-    writer = pd.ExcelWriter(os.path.join(directory,'joined_codes.xlsx'), engine='xlsxwriter')
-    for code in df_npis_combinations.keys():
-        df_npis_combinations[code][1].to_excel(writer, sheet_name=code)
+                for column_other in range(column):
+                    df_count[maincode][1].iloc[column, column_other] += len(set(
```
[the remark this answers is not in the document]
Comment (patricklnz) [reply pycode/memilio-epidata/memilio/epidata/getNPIData.py]: Also for every county.

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py (hunk 8 above)
Comment (patricklnz) [reply pycode/memilio-epidata/memilio/epidata/getNPIData.py:1059]: we can use filter for both cases

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py (hunk 7 above)
Comment (patricklnz) [reply pycode/memilio-epidata/memilio/epidata/getNPIData.py:720]: yes

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 12]
```diff
@@ -529,32 +532,34 @@ def get_npi_data(fine_resolution=2,
...
         # drop 0 column if existent
         try:
             df_npis_combinations_pre.drop(columns=0)
         except KeyError:
             pass
         # rename essential columns and throw away others
-        columns_combinations = np.where((df_npis_combinations_pre=='x').any()==True)[0] #maybe rename columns_used ?
-        column_names = ['Unnamed: ' + str(i) for i in range(columns_combinations[0], columns_combinations[-1]+1)]
+        columns_combinations = np.where((df_npis_combinations_pre == 'x').any() == True)[
+            0]  # maybe rename columns_used ?
```
[the remark this answers is not in the document]
Comment (patricklnz) [reply pycode/memilio-epidata/memilio/epidata/getNPIData.py:560]: changed it.

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py (hunk 8 above)
Comment (patricklnz) [reply pycode/memilio-epidata/memilio/epidata/getNPIData.py:1060]: Sorry, not exactly true. We can use filter if the npi code is either the column or the index. Else it is easier to use str.contains. Also `df.filter(npi_code)` returns a dataframe and `df.columns.str.contains(regex=filter)` returns an array.

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py (hunk 10 above)
Comment (patricklnz) [reply pycode/memilio-epidata/memilio/epidata/getNPIData.py:963]: Defined 2 new variables

**Review by mknaranja — COMMENTED** [review]
Hi @patricklnz & @annawendler,

I went through the all your changes since April and reduced some of the redundant functionality. I hope that I did not introduce new errors. Please check this.

Furthermore, the tests would need to be adapted as I renamed and reduced the number of introduced functions. Could you take over here?

**Review by mknaranja — COMMENTED** [review]
Looks good so far. Have you finished here @patricklnz and @annawendler ? I just saw there is one final comment on the multiple zeros. Can you answer that?

I just merged the main into the branch.

File: pycode/memilio-epidata/memilio/epidata/getNPIData.py [hunk 13]
```diff
@@ -566,25 +587,26 @@ def get_npi_data(fine_resolution=2,
...
+                    len(df_npis_combinations_pre['Variablenname'][npi_groups_idx[i]]))},
                 np.eye(len(npi_groups_idx[i]))]
             for i in range(len(npi_groups_combinations_unique))}
 
-
         # run through all groups and set possible combinations according to
         # read combination matrix
-        start_comb_matrix = list(df_npis_combinations_pre.columns).index('Variablenname')+3
+        # find begin of combination matrix
+        # there may be multiple columns named '0', so find first '1' column
```
[the remark this answers is not in the document]
Comment (annawendler) [reply pycode/memilio-epidata/memilio/epidata/getNPIData.py:641]: I think this is from when the 'Massnahmenindex' was new and the column wasn't named yet. I don't think this is necessary anymore and I would change it in the next commit to look for the first column named '0'. Can you confirm this @patricklnz ?

**Review by patricklnz — APPROVED** [review]
Approved!

**Pull request merged**